### PR TITLE
Top k processor

### PR DESCRIPTION
**Pull request by mirath**
This processor objective is to provide the possibility to pick the top k metrics from a set of metrics. It aims to solve #3192. It also solves #3562, as that issue was preventing the test to run succesfully

The pluging uses:

- Selectors to pick the metrics that will be filtered
- A group by clause to group the metrics based on metric names and tags
- A list of fields
- An aggregator function to agreggate each field of the grouped metrics
- Once the aggregations for each group are calculated, the plugin orders the groups by each aggregation and returns all the metrics of each group in the top k spots

PR #3400 closed in favor of this one, as I mangled the branch through several irresponsible rebases

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

**Comment by mirath** [comment]
@danielnelson Here is the new PR. Much better XD

File: plugins/processors/topk/topk.go [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+  # Metrics are grouped based on their tags and name. The plugin aggregates the selected fields of
+  # these groups of metrics and sorts the groups based these aggregations
+  group_by = ["process_name"]  # Over which tags should the aggregation be done. Default: []
+  group_by_metric_name = false # Wheter or not to also group by metric name. Default: false
+
+  # The plugin can aggregate over several fields. If more than one field is specified, an aggregation is calculated per group per field
+  # The plugin returns a metric if it is in a group in the top k groups ordered by any of the aggregations of the selected fields
+  # This effectively means that more than K metrics may be returned. If you need to return only the top k metrics regardless of grouping, use the simple_topk setting
+  fields = ["memory_rss"]      # Over which fields are the top k are calculated. Default: ["value"]
+  aggregation = "avg"          # What aggregation to use. Default: "avg". Options: sum, avg, min, max
```
Comment (danielnelson) [inline plugins/processors/topk/topk.go]: Let's call `avg` -> `mean` to be consistent with basicstats aggregator

Comment (danielnelson) [inline plugins/processors/topk/topk.go]: I wonder if we need to have support for different aggregations per field.

File: plugins/processors/topk/topk.go [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+}
+
+var sampleConfig = `
+[[processors.topk]]
+  period = 10                  # How many seconds between aggregations. Default: 10
+  k = 10                       # How many top metrics to return. Default: 10
+
+  # Metrics are grouped based on their tags and name. The plugin aggregates the selected fields of
+  # these groups of metrics and sorts the groups based these aggregations
+  group_by = ["process_name"]  # Over which tags should the aggregation be done. Default: []
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: If you have metrics with different tag sets that you want to aggregate together. For example, I want the top 10 processes by RSS size across all the instances in the cluster0.

Are you thinking that there should be a way to say "I want to aggregate across all tags" without having to specify tag by tag? That would certainly help in environments with dynamic tags

File: plugins/processors/topk/topk.go (hunk 1 above)
Comment (mirath) [reply plugins/processors/topk/topk.go]: That is a very good question. I thought about it and decided that someone down the line could contribute such a feature. Do you think is a valuable enough capability to delay the release and work on it before merging?

File: plugins/processors/topk/topk.go [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+                               # Useful for debugging. Default: ""
+
+  position_field = ""          # This settings provides a way to know the position of each metric in the top k.
+                               # If set to a value different than "", then a field (which name will be prefixed with the value of this setting) will be
+                               # added to each every metric for each field over which an aggregation was made.
+                               # This field will contain the ranking of the group that the metric belonged to. When aggregating
+                               # over several fields, several fields will be added (one for each field over which an aggregation was calculated).
+                               # Useful for debugging. Default: ""
+
+  aggregation_field = ""       # This setting provies a way know the what values the plugin is generating when aggregating the fields
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: I like this idea. However, the reason I let the user specify the name of the field, is to give a way to handle the edge case where they already have a field named the field I'm adding.

We could, aside from using the `add_field_aggregate` and `add_field_rank` settings. Add settings with sensible default values to override the prefix used for the added fields, something like `field_aggreate_prefix="aggregation_"` and `field_rank_prefix="rank_"`

File: plugins/processors/topk/topk_test.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	for _, ag := range aggregators {
+		topk.Aggregation = ag
+
+		//The answer is equal to the original set for these particual scenarios
+		runAndCompare(&topk, MetricsSet1, MetricsSet1, "SmokeAggregator_"+ag, t)
+	}
+}
+
+// AggregationField + Avg aggregator
+func TestTopkAvgAggregationField(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk_test.go]: I'll try to improve them. They were hard to follow even while I was making them.

File: plugins/processors/topk/topk.go [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+
+		t.Reset()
+
+		return ret
+	}
+
+	return []telegraf.Metric{}
+}
+
+func min(a, b int) int {
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: It's called in line 221

    // Create a one dimentional list with the top K metrics of each key
    for i, ag := range aggregations[0:min(t.K, len(aggregations))] {

File: plugins/processors/topk/topk.go [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+	groupbykey string
+	values     map[string]float64
+}
+
+type Aggregations struct {
+	metrics []MetricAggregation
+	field   string
+}
+
+func (ags Aggregations) Len() int {
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: I can't. The sort interface asks for a direct value:
```
plugins/processors/topk/topk.go:165: cannot use aggs (type Aggregations) as type sort.Interface in argument to sort.Sort:
	Aggregations does not implement sort.Interface (Len method has pointer receiver)
plugins/processors/topk/topk.go:167: cannot use aggs (type Aggregations) as type sort.Interface in argument to sort.Reverse:
	Aggregations does not implement sort.Interface (Len method has pointer receiver)
```

File: plugins/processors/topk/topk.go [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+  group_by_metric_name = false # Wheter or not to also group by metric name. Default: false
+
+  # The plugin can aggregate over several fields. If more than one field is specified, an aggregation is calculated per group per field
+  # The plugin returns a metric if it is in a group in the top k groups ordered by any of the aggregations of the selected fields
+  # This effectively means that more than K metrics may be returned. If you need to return only the top k metrics regardless of grouping, use the simple_topk setting
+  fields = ["memory_rss"]      # Over which fields are the top k are calculated. Default: ["value"]
+  aggregation = "avg"          # What aggregation to use. Default: "avg". Options: sum, avg, min, max
+
+  bottomk = false              # Instead of the top k largest metrics, return the bottom k lowest metrics. Default: false
+  simple_topk = false          # If true, this will override any GroupBy options and assign each metric its own individual group. Default: false
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: I think we should treat an empty `group_by` to aggregate over all the tags of the metric

File: plugins/processors/topk/topk.go [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+	return topk
+}
+
+func NewTopKProcessor() telegraf.Processor {
+	topk := NewTopK()
+	return &topk
+}
+
+var sampleConfig = `
+[[processors.topk]]
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: @danielnelson I've corrected this and several other issues. I've not pushed because in fixing this issue I changed lines tied to several open discussions, and I don't want them to show as updated when they are not yet resolved. When we come to a conclusion regarding the outstanding comments I will push my latest changes.

File: plugins/processors/topk/topk.go [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+  # This effectively means that more than K metrics may be returned. If you need to return only the top k metrics regardless of grouping, use the simple_topk setting
+  fields = ["memory_rss"]      # Over which fields are the top k are calculated. Default: ["value"]
+  aggregation = "avg"          # What aggregation to use. Default: "avg". Options: sum, avg, min, max
+
+  bottomk = false              # Instead of the top k largest metrics, return the bottom k lowest metrics. Default: false
+  simple_topk = false          # If true, this will override any GroupBy options and assign each metric its own individual group. Default: false
+  drop_no_group = true         # Drop any metrics that do fit in any group (due to nonexistent tags). Default: true
+  drop_non_top = true          # Drop the metrics that do not make the cut for the top k. Default: true
+
+  group_by_tag = ""            # The plugin assigns each metric a GroupBy tag generated from its name and tags.
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: Solved

File: plugins/processors/topk/topk.go [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+	for _, m := range in {
+		t.groupBy(m)
+	}
+
+	// If enough time has passed
+	elapsed := time.Since(t.lastAggregation)
+	if elapsed >= time.Second*time.Duration(t.Period) {
+		// Generate aggregations list using the selected fields
+		aggregations := make([]MetricAggregation, 0, 100)
+		var aggregator func([]telegraf.Metric, []string) map[string]float64 = t.getAggregationFunction(t.Aggregation)
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: Solved

File: plugins/processors/topk/topk.go [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+				if aggField != "" || posField != "" || groupTag != "" {
+					for _, m := range t.cache[ag.groupbykey] {
+						if aggField != "" && m.HasField(field) {
+							m.AddField(aggField+"_"+field, ag.values[field])
+						}
+						if posField != "" {
+							m.AddField(posField+"_"+field, i+1) //+1 to it starts from 1
+						}
+						if groupTag != "" {
+							m.AddTag(groupTag, ag.groupbykey) //+1 to it starts from 1
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: Solved

File: plugins/processors/topk/topk.go [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+	case float64:
+		return v, true
+	case int64:
+		return float64(v), true
+	default:
+		return 0, false
+	}
+}
+
+func init() {
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: Solved

File: plugins/processors/topk/topk.go (hunk 1 above)
Comment (mirath) [reply plugins/processors/topk/topk.go]: Solved

File: plugins/processors/topk/topk.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: Solved

File: plugins/processors/topk/topk.go (hunk 1 above)
Comment (danielnelson) [reply plugins/processors/topk/topk.go]: I think we can add this in later.

File: plugins/processors/topk/topk.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/topk.go]: Yes and aggregate across all tags should probably be the default, I would do what you mentioned with empty list matching all.  Use the `filter.Filter` type which support globs (not regex), which will make it behave like the measurement filtering options.

File: plugins/processors/topk/topk.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/topk.go]: That's a good idea, but I think we should use a suffix mostly for consistency with existing aggregation style.

File: plugins/processors/topk/topk.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/topk.go]: Okay, BTW I have been heard from my colleagues that it is now preferred to use `sort.SliceStable` with a less func instead of defining ordering on the type.

**Comment by mirath** [comment]
@danielnelson I finished addressing the issues you brought up. I also made `group_by_metric_name` true by default, as I figure that that way the processor is most likely to work out of the box

Let me know if you have any more comments, or if we are ready to merge (and if the plugin would be available for Telegraf 1.6)

**Comment by danielnelson** [comment]
I'll try to take a look as soon as I can, but this will have to be a 1.7 addition, sorry.

File: plugins/processors/topk/topk.go [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+	// If enough time has passed
+	elapsed := time.Since(t.lastAggregation)
+	if elapsed >= time.Second*time.Duration(t.Period) {
+		// Generate aggregations list using the selected fields
+		aggregations := make([]MetricAggregation, 0, 100)
+		aggregator, err := t.getAggregationFunction(t.Aggregation)
+		if err != nil {
+			// If we could not generate the aggregation
+			// function, fail hard by dropping all metrics
+			log.Print(err)
```
Comment (danielnelson) [inline plugins/processors/topk/topk.go]: `E! [processors.topk]: %v`

File: plugins/processors/topk/topk.go [hunk 14]
```diff
@@ -0,0 +1,468 @@
...
+				keys = append(keys, tag+"="+value+"&")
+			}
+		}
+		// Sorting the selected tags is necessary because dictionaries
+		// do not ensure any specific or deterministic ordering
+		sort.SliceStable(keys, func(i, j int) bool { return keys[i] < keys[j] })
+		for _, str := range keys {
+			groupkey += str
+		}
+	}
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: But wouldn't this remove information from the metric that maybe useful down the pipeline for the user. It seems weird to me that the user, for example, groups cpu metrics by usage and loses information about the datacenter that created each metric, which could be useful when diagnosing a but (ie. You see a spike in cpu and 70% of the metrics in the spike came from datacenter A)

File: plugins/processors/topk/README.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+  ## If true, this will override any GroupBy options and assign each metric
+  ## its own individual group. Default: false
+  # simple_topk = false
+
+  ## Drop any metrics that do fit in any group (due to nonexistent tags)
+  # drop_no_group = true
+
+  ## Drop the metrics that do not make the cut for the top k
+  # drop_non_top = true
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: I don't see how `k` implements this functionality. This is here to give the user the option *not to drop any metrics* (by setting this to false), which might seem like a dumb option in a plugin whose sole purpose is to keep a subset of the metrics, but it can be useful because this plugin can also *annotate* metrics.

This option is here for users that would like TopK to _annotate_ their metrics (as you can do with the rank and aggregate field options) but not drop them, letting them effectively do some of the dropping or any other post processing afterwards.

I do think though that that is a very edgy case, and that for the vast mayority of users this will remain as the default, so I'm not terribly attached to it. If you guys still want it gone, Ill delete the feature

File: plugins/processors/topk/README.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+  # k = 10
+
+  ## Metrics are grouped based on their tags and name. The plugin aggregates
+  ## the selected fields of these groups of metrics and sorts the groups based
+  ## these aggregations
+
+  ## Over which tags should the aggregation be done. Globs can be specified, in
+  ## which case any tag matching the glob will aggregated over. If set to an
+  ## empty list is no aggregation over tags is done
+  # group_by = ['*']
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: No, `[]` will just ignore every tag. In contrast `["*"]` will group by all tags

PS. We should have a centralized discussion about the meanings of `[]`, `[""]` and `["*"]`. What I propose is:

* `[]`: Ignore tags
* `[""]`: Do a simple topk, putting each metric in it's own bucket indivually, regardless of tags
* `["*"]`: Group by all tags

File: plugins/processors/topk/README.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+  # group_by_metric_name = true
+
+  ## The plugin can aggregate over several fields. If more than one field is
+  ## specified, an aggregation is calculated per group per field.
+
+  ## The plugin returns a metric if it's in a group in the top k groups,
+  ## ordered by any of the aggregations of the selected fields
+
+  ## This effectively means that more than K metrics may be returned. If you
+  ## need to return only the top k metrics regardless of grouping, use the simple_topk setting
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: The processor goes through these steps:

1. Groups metrics in buckets using their tags & metrics name as key
2. Aggregates each of the selected fields for each bucket by some aggregation function (sum, mean, etc)
3. Orders the **buckets** by one of the generated aggregations, returns **all metrics in the top k** buckets, then reorders the buckets by the next of the generated aggregations, returns those metrics, etc, etc, etc, until it runs out of fields.
4. Finish the execution

So if you group by `PID` and aggregate over `cpu` and `memory` with a `K` of 10, you could get between 10 and 20 processes. This is also because PID's are unique, if I group by `datacenter` or `OS`, I could get **more** than 20 processes

This gives "complete" and "intuitive" results. What deficiencies do you see that make you look for an alternative?

File: plugins/processors/topk/topk.go [hunk 18]
```diff
@@ -0,0 +1,468 @@
...
+	return groupkey, nil
+}
+
+func (t *TopK) groupBy(m telegraf.Metric) {
+	// Generate the metric group key
+	groupkey, err := t.generateGroupByKey(m)
+	if err != nil {
+		// If we could not generate the groupkey, fail hard
+		// by dropping this and all subsequent metrics
+		log.Print(err)
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: Agreed

File: plugins/processors/topk/README.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+  # drop_no_group = true
+
+  ## Drop the metrics that do not make the cut for the top k
+  # drop_non_top = true
+
+  ## The plugin assigns each metric a GroupBy tag generated from its name and
+  ## tags. If this setting is different than "" the plugin will add a
+  ## tag (which name will be the value of this setting) to each metric with
+  ## the value of the calculated GroupBy tag. Useful for debugging
+  # group_by_tag = ""
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: Agreed

File: plugins/processors/topk/topk.go [hunk 20]
```diff
@@ -0,0 +1,468 @@
...
+	var ret []telegraf.Metric = make([]telegraf.Metric, 0, 0)
+	for _, m := range in {
+		t.groupBy(m)
+	}
+
+	// If enough time has passed
+	elapsed := time.Since(t.lastAggregation)
+	if elapsed >= time.Second*time.Duration(t.Period) {
+		// Generate aggregations list using the selected fields
+		aggregations := make([]MetricAggregation, 0, 100)
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: Ok

File: plugins/processors/topk/topk.go [hunk 21]
```diff
@@ -0,0 +1,468 @@
...
+	if t.SimpleTopk {
+		return strconv.FormatUint(m.HashID(), 16), nil
+	}
+
+	// Create the filter.Filter objects if they have not been created
+	if t.tagsGlobs == nil && len(t.GroupBy) > 0 {
+		var err error
+		t.tagsGlobs, err = filter.Compile(t.GroupBy)
+		if err != nil {
+			log.Printf("Could not complile tags globs: %s", t.GroupBy)
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: Ok

File: plugins/processors/topk/README.md [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+
+  ## These settings provide a way to know the position of each metric in
+  ## the top k. The 'add_rank_field' setting allows to specify for which
+  ## fields the position is required. If the list is non empty, then a field
+  ## will be added to each every metric for each field present in the
+  ## 'add_rank_field'. This field will contain the ranking of the group that
+  ## the metric belonged to when aggregated over that field.
+  ## The name of the field will be set to the name of the aggregation field,
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: Yeah, and if you aggregate by many fields (lets say: sum, mean and max), you would get a sum_rank, mean_rank and max_rank on each processed metric

File: plugins/processors/topk/README.md [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+### Tags:
+
+This processor does not add tags by default. But the setting `group_by_tag` will add a tag if set to anything other than ""
+
+
+### Fields:
+
+This processor does not add fields by default. But the settings `position_field` and `aggregation_field` will add one or several fields if set to anything other than ""
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: ok

File: plugins/processors/topk/README.md [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+This plugin groups the metrics based on their name and tags, then generates aggregates values across each group base on fields selected by the user. It then sorts these groups based on these aggregations and returns any metric that belongs to a group in the top k (sorted by any of the aggregations). This means that when calculating the top k, more than k metrics may be returned.
+
+If only the very top k metrics are needed, regardless of grouping, the simple_topk setting will force each metric into its own individual group
+
+### Configuration:
+
+```toml
+[[processors.topk]]
+  ## How many seconds between aggregations
+  # period = 10
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: Agreed

File: plugins/processors/topk/topk.go [hunk 25]
```diff
@@ -0,0 +1,468 @@
...
+		return b
+	}
+	return a
+}
+
+func convert(in interface{}) (float64, bool) {
+	switch v := in.(type) {
+	case float64:
+		return v, true
+	case int64:
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: Cool, will do

File: plugins/processors/topk/topk.go (hunk 13 above)
Comment (mirath) [reply plugins/processors/topk/topk.go]: Ok

File: metric/metric.go [hunk 26]
```diff
@@ -508,20 +508,51 @@ func (m *metric) RemoveField(key string) error {
 }
 
 func (m *metric) Copy() telegraf.Metric {
-	return copyWith(m.name, m.tags, m.fields, m.t)
+	return duplicateMetricWith(m, nil, nil, nil, nil)
 }
 
-func copyWith(name, tags, fields, t []byte) telegraf.Metric {
+func duplicateMetricWith(m *metric, name, tags, fields, t []byte) telegraf.Metric {
```
[the remark this answers is not in the document]
Comment (mirath) [reply metric/metric.go]: Let me check how to do that without contaminating the diff like last time.

File: plugins/processors/topk/TODO [hunk 27]
```diff
@@ -0,0 +1,9 @@
+Clarify tests
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/TODO]: ok

File: plugins/processors/topk/README.md [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+  ## the selected fields of these groups of metrics and sorts the groups based
+  ## these aggregations
+
+  ## Over which tags should the aggregation be done. Globs can be specified, in
+  ## which case any tag matching the glob will aggregated over. If set to an
+  ## empty list is no aggregation over tags is done
+  # group_by = ['*']
+
+  ## Wheter or not to also group by metric name
+  # group_by_metric_name = true
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: I'm trying to think of a case were this would hurt us, because I feel that it will. But I can't think on anything (though this will force the user to group by some tag, otherwise the plugin will never drop anything)

If I think about an edge case I'll let you know. In the mean time I'll remove the flag

File: plugins/processors/topk/README.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: This was a documentation error in the README. The real option is called `add_groupby_tag`. Is that name acceptable?

File: plugins/processors/topk/topk.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: Ps. This is logged in the groupBy function. I applied the change that you suggested, and it will return the error to this block:
```
	// Generate the metric group key
	groupkey, err := t.generateGroupByKey(m)
	if err != nil {
		// If we could not generate the groupkey, fail hard
		// by dropping this and all subsequent metrics
		log.Print("E! [processors.topk]: could not generate group key: %v", err)
		return
	}
```

Which will just log and drop the metrics.

File: plugins/processors/topk/topk.go [hunk 29]
```diff
@@ -0,0 +1,468 @@
...
+
+						// Add the rank relative to the current field if requested
+						_, addRankField := t.rankFieldSet[field]
+						if addRankField && m.HasField(field) {
+							m.AddField(field+rankFieldSuffix, i+1)
+						}
+
+						// Add the generated groupby key if requested
+						if groupTag != "" {
+							m.AddTag(groupTag, ag.groupbykey)
```
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: There wasn't a good place to put this later, but I was able to place it earlier, when the groupby key is computed for the very first time. This way this is only done once per metric

File: plugins/processors/topk/README.md [hunk 30]
```diff
@@ -0,0 +1,97 @@
...
+
+  ## What aggregation to use. Options: sum, mean, min, max
+  # aggregation = "mean"
+
+  ## Instead of the top k largest metrics, return the bottom k lowest metrics
+  # bottomk = false
+
+  ## If true, this will override any GroupBy options and assign each metric
+  ## its own individual group. Default: false
+  # simple_topk = false
```
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/README.md]: I think the confusion is that I'm thinking of `group_by` having the opposite effect.  In my mind when we `group by *` we are creating buckets for all tag combinations, not grouping together.  I expect it to work this way because of InfluxDB:

```
$ influx
> insert foo,a=x value=42
> insert foo,a=y value=42
> insert foo,a=z value=42
```

If I don't group they are the same series:
```
> select * from foo
name: foo
time                a value
----                - -----
1524523227856932634 x 42
1524523235152945692 y 42
1524523238021967153 z 42
```

If I group by * then I get one series per tag value:
```
> select * from foo group by *
name: foo
tags: a=x
time                value
----                -----
1524523227856932634 42

name: foo
tags: a=y
time                value
----                -----
1524523235152945692 42

name: foo
tags: a=z
time                value
----                -----
1524523238021967153 42
```

File: plugins/processors/topk/README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/README.md]: I think this is reasonable but there may be some confusion over why they are getting more than `k` results.

File: plugins/processors/topk/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/README.md]: If `k` was set to a suitably large value then you would get all values, which is why I feel like the options are related.  In some contexts it is common for 0 to indicate infinity, such as when specifying a timeout for a network operation.  Either way is okay.

File: plugins/processors/topk/README.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: I expressed myself wrong. I should've say "group by "*" would group metrics that have the same tags with the same values together".

Basically, for grouping, given a list of tags the plugin creates a string of the form `metric_name&tag1=val1&tag2=val2...` for each metric. If a metric doesn't have a tag it doesn't get added to the string, it then places the metric in a map with its generated string as a key.

[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: So it actually behaves as you would expect

File: plugins/processors/topk/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: Yeah, but a very large k would also **annotate** all values, which it's something you might not want.

File: plugins/processors/topk/README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: Yeah, but the think about the problem is that unless you are doing a "simple" top k, you don't know before hand how many points of data you are gonna get.

Like if I made it so it only gives you the very top k of all the ingested metrics, some of the points could come in and out very rapidly, which would in a very unuseful graph. Maybe we just need make it clear in the documentation or change the name of the plugin.

File: plugins/processors/topk/README.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/README.md]: Does that mean that by default you would want `group_by = ["*"]` and that would be identical to `simple_topk`?

File: plugins/processors/topk/README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/README.md]: Okay, nevermind about this comment, I think how you have it is the best.

File: plugins/processors/topk/README.md [hunk 31]
```diff
@@ -0,0 +1,97 @@
...
+
+  ## These settings provide a way to know the position of each metric in
+  ## the top k. The 'add_rank_field' setting allows to specify for which
+  ## fields the position is required. If the list is non empty, then a field
+  ## will be added to each every metric for each field present in the
+  ## 'add_rank_field'. This field will contain the ranking of the group that
+  ## the metric belonged to when aggregated over that field.
+  ## The name of the field will be set to the name of the aggregation field,
+  ## suffixed by the value of the 'rank_field_suffix' setting
+  # add_rank_field = []
```
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/README.md]: I did search the Telegraf repo for '_rank' and came up with no hits, so I'm not too concerned about conflicts.  My opinion is that completely avoiding them is more trouble than it is worth in complexity or potential ugliness.

I will leave it up to you if you think we should keep it configurable, but don't try to further scope it.  Keep in mind we will likely add a processor that can rename fields too.

File: plugins/processors/topk/README.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: By default `group_by = ["*"]` its what most people would want I think. But its not the same as `simple_topk`.

The thing to notice is that for `group_by = ["*"]` these metrics belong in the same bucket:
```
foo,a=x, b=y value=42
foo,a=x, b=y value=690
```

But for `simple_topk` they belong in separate buckets. The `simple_topk` option puts every metric in its own bucket **irregladless of tags**, effectively **making `N` buckets each containing a single metric** (where N is the number of ingested metrics)

[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/README.md]: I don't think is is correct, the HashID() is based on measurement name + tagset, they would both have the same ID and they would get aggregated together.

[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/README.md]: No, I feel like it does not behave well within this plugin, as the aggregation aspect is effectively disabled and if you have a period that is different from the input plugin interval you end up with a data set that may be missing some of the intervals completely.

Maybe I shouldn't mention it, but I do wonder if this plugin could be decomposed into several plugins.  The aggregation aspect could be farmed out to the basicstats plugin completely, and then there could be a separate rank aggregator that only ranks based on fields.  Finally there is a processor that filters based on field values (the rank).
```
inputs -> basicstats (or other aggregators) -> ranker aggregator  -> topk processor
```

One thing I like about this is that it separates out the aggregator and processor aspects, so that the plugin is no longer a processor with aggregator like features.  Also if as we add new aggregators you can use them automatically.  Finally, it seems more straightforward to use as a pure annotator, you just leave out the top processor piece.

[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: Hi, sorry for the delay, I've had a hectic few days, and also your comment merited some extended thought.

I checked the basicstats plugin and it doesn't have any grouping feature, so I'm guessing that you mean we could port the grouping feature here to basicstats, make a new ranker aggregator, and leave this processor to solely do the filtering.

I like that approach, and I think is better engineered, but I feel that it would delay the plugin an unnecessary amount of time (the old lets-not-release-the-feature-until-is-perfect syndrome)

What I could do is refactor some of the code to "divide it". So I would have a library that would generate the basic stats based on grouping, then another that would generate the ranks, and the processor itself that would do the filtering.

It will delay the release slightly, but that way factoring out things in the future would be significantly simpler, and we get to push this value onto the users sooner. I think is a nice half way to get our cake and eat it too.

What do you think?

[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/README.md]: I think we should probably shelve the idea to split it out, its interesting but I think we shouldn't do it now.  I'm not 100% sure the design would work out, and even if it does I think it would be trickier to setup since you would need to combine so many pieces.  On top of that it would take a fair bit of time to refactor everything.

Let's continue with the current structure on this plugin, and in the future we can experiment more with the modular idea.

File: plugins/processors/topk/topk.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/topk.go]: @danielnelson Still waiting on input on this comment

File: plugins/processors/topk/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (mirath) [reply plugins/processors/topk/README.md]: @danielnelson Still waiting on input on this comment. Going by our previous discussion regarding simple top k (which since then I've already removed) I think it makes sense to remove this option too. It's an instance of the plugin trying to do too many things.

File: plugins/processors/topk/topk.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/topk.go]: I think you are right, we don't want to remove any tags here.

File: plugins/processors/topk/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/processors/topk/README.md]: Yeah, I think I would remove the option, wanting to annotate some of the values seems like a fringe use.

**Comment by mirath** [comment]
I've addressed almost all the issues you brought up. I just need to rebase to drop the changes in metrics.go.

However I realized there was a case we were not handling: what to do with a metric that does not have any of the fields over which we are aggregating.

It seems like a fringe case, the options as I see them are:
  1.  "Let them be", not count them in the computations, but let them go through untouched.
  2.  Drop them

My intuition would be to "let them be" (I don't like to kill things that I didn't generate), but given the use cases of telegraf, it seems like a fringe case, so we could just drop them.

**Comment by danielnelson** [comment]
I think we should probably drop items that don't have any fields we are aggregating, since they won't be "top" (or even bottom).

**Comment by mirath** [comment]
So I messed up the rebase again X.X

I'm closing this PR and opening a new one.
